### PR TITLE
Add Compaction to Channelz Registry

### PR DESCRIPTION
**Pull request by ncteisen**
Adds a compaction operation if the registry is 1/3 empty.

This protects the registry from growing unbounded and not cleaning up after entities are deleted.

Tracking: #15340

### CODE REVIEW

**Comment by ncteisen** [comment]
I plan to add more tests to channelz_registry_test and inline_vector_test later tonight.

Just wanted to put this out now and response to some of my other PRs

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                         FILE SIZE
 ++++++++++++++ GROWING                                                           ++++++++++++++
  +0.0%    +344 [None]                                                            +7.27Ki  +0.1%
      +0.0%    +312 [Unmapped]                                                        +7.27Ki  +0.1%
      +3.1%     +32 [None]                                                                  0  [ = ]
   +20%    +512 src/core/lib/channel/channelz_registry.cc                            +512   +20%
      [NEW]    +214 grpc_core::channelz::ChannelzRegistry::MaybePerformCompaction        +214  [NEW]
      [NEW]    +150 grpc_core::channelz::ChannelzRegistry::FindByUuid                    +150  [NEW]
       +47%     +80 grpc_core::channelz::ChannelzRegistry::InternalUnregister             +80   +47%
       +55%     +21 grpc_core::channelz::ChannelzRegistry::ChannelzRegistry               +21   +55%
       +12%     +18 [Unmapped]                                                            +18   +12%
      +7.4%     +16 grpc_core::channelz::ChannelzRegistry::InternalRegister               +16  +7.4%
       +15%     +13 grpc_core::channelz::ChannelzRegistry::InternalGet                    +13   +15%

  +0.1%    +856 TOTAL                                                             +7.77Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +32  +0.0%

  [ = ]       0 TOTAL      +32  +0.0%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,965,517      Total (>)      1,964,892

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,888,195      Total (>)     10,887,719

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                         FILE SIZE
 ++++++++++++++ GROWING                                                           ++++++++++++++
  +0.0%    +344 [None]                                                            +3.30Ki  +0.0%
      +0.0%    +312 [Unmapped]                                                        +3.30Ki  +0.0%
      +3.1%     +32 [None]                                                                  0  [ = ]
   +20%    +512 src/core/lib/channel/channelz_registry.cc                            +512   +20%
      [NEW]    +214 grpc_core::channelz::ChannelzRegistry::MaybePerformCompaction        +214  [NEW]
      [NEW]    +150 grpc_core::channelz::ChannelzRegistry::FindByUuid                    +150  [NEW]
       +47%     +80 grpc_core::channelz::ChannelzRegistry::InternalUnregister             +80   +47%
       +55%     +21 grpc_core::channelz::ChannelzRegistry::ChannelzRegistry               +21   +55%
       +12%     +18 [Unmapped]                                                            +18   +12%
      +7.4%     +16 grpc_core::channelz::ChannelzRegistry::InternalRegister               +16  +7.4%
       +15%     +13 grpc_core::channelz::ChannelzRegistry::InternalGet                    +13   +15%

  +0.1%    +856 TOTAL                                                             +3.80Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +32  +0.0%

  [ = ]       0 TOTAL      +32  +0.0%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,965,541      Total (>)      1,964,916

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,888,196      Total (>)     10,887,712

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                         FILE SIZE
 ++++++++++++++ GROWING                                                           ++++++++++++++
  +0.0%    +344 [None]                                                            +3.30Ki  +0.0%
      +0.0%    +312 [Unmapped]                                                        +3.30Ki  +0.0%
      +3.1%     +32 [None]                                                                  0  [ = ]
   +20%    +512 src/core/lib/channel/channelz_registry.cc                            +512   +20%
      [NEW]    +214 grpc_core::channelz::ChannelzRegistry::MaybePerformCompaction        +214  [NEW]
      [NEW]    +150 grpc_core::channelz::ChannelzRegistry::FindByUuid                    +150  [NEW]
       +47%     +80 grpc_core::channelz::ChannelzRegistry::InternalUnregister             +80   +47%
       +55%     +21 grpc_core::channelz::ChannelzRegistry::ChannelzRegistry               +21   +55%
       +12%     +18 [Unmapped]                                                            +18   +12%
      +7.4%     +16 grpc_core::channelz::ChannelzRegistry::InternalRegister               +16  +7.4%
       +15%     +13 grpc_core::channelz::ChannelzRegistry::InternalGet                    +13   +15%

  +0.1%    +856 TOTAL                                                             +3.80Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +32  +0.0%

  [ = ]       0 TOTAL      +32  +0.0%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                         FILE SIZE
 ++++++++++++++ GROWING                                                           ++++++++++++++
  +0.0%    +344 [None]                                                            +7.25Ki  +0.1%
      +0.0%    +312 [Unmapped]                                                        +7.25Ki  +0.1%
      +3.1%     +32 [None]                                                                  0  [ = ]
   +20%    +512 src/core/lib/channel/channelz_registry.cc                            +512   +20%
      [NEW]    +214 grpc_core::channelz::ChannelzRegistry::MaybePerformCompaction        +214  [NEW]
      [NEW]    +153 grpc_core::channelz::ChannelzRegistry::FindByUuid                    +153  [NEW]
       +47%     +80 grpc_core::channelz::ChannelzRegistry::InternalUnregister             +80   +47%
       +55%     +21 grpc_core::channelz::ChannelzRegistry::ChannelzRegistry               +21   +55%
      +7.4%     +16 grpc_core::channelz::ChannelzRegistry::InternalRegister               +16  +7.4%
      +9.9%     +15 [Unmapped]                                                            +15  +9.9%
       +15%     +13 grpc_core::channelz::ChannelzRegistry::InternalGet                    +13   +15%

  +0.1%    +856 TOTAL                                                             +7.75Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +32  +0.0%

  [ = ]       0 TOTAL      +32  +0.0%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,965,405      Total (>)      1,964,816

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,531      Total (>)     10,887,064

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: src/core/lib/channel/channelz_registry.cc [hunk 1]
```diff
@@ -56,23 +56,71 @@ ChannelzRegistry::~ChannelzRegistry() { gpr_mu_destroy(&mu_); }
 intptr_t ChannelzRegistry::InternalRegister(BaseNode* node) {
   MutexLock lock(&mu_);
   entities_.push_back(node);
-  intptr_t uuid = entities_.size();
-  return uuid;
+  return ++uuid_generator_;
+}
+
+void ChannelzRegistry::MaybePerformCompaction() {
```
Comment (markdroth) [inline src/core/lib/channel/channelz_registry.cc]: Please add a `Locked` suffix to the name.

File: src/core/lib/channel/channelz_registry.cc [hunk 2]
```diff
@@ -56,23 +56,71 @@ ChannelzRegistry::~ChannelzRegistry() { gpr_mu_destroy(&mu_); }
...
+      }
+    }
+    for (int i = 0; i < num_empty_slots_; ++i) {
+      entities_.pop_back();
+    }
+    num_empty_slots_ = 0;
+  }
+}
+
+int ChannelzRegistry::FindByUuid(intptr_t target_uuid) {
```
Comment (markdroth) [inline src/core/lib/channel/channelz_registry.cc]: Please add a `Locked` suffix to the name.

File: test/core/channel/channelz_registry_test.cc [hunk 3]
```diff
@@ -43,56 +43,151 @@ namespace grpc_core {
...
+class ChannelzRegistryTest : public ::testing::Test {
+ protected:
+  // ensure we always have a fresh registry for tests.
+  void SetUp() override { ChannelzRegistry::Init(); }
+
+  void TearDown() override { ChannelzRegistry::Shutdown(); }
+};
+
+TEST_F(ChannelzRegistryTest, UuidStartsAboveZeroTest) {
+  UniquePtr<BaseNode> channelz_channel(
```
Comment (markdroth) [inline test/core/channel/channelz_registry_test.cc]: Could use `MakeUnique<>` here.

Same thing in all of these tests.

File: test/core/channel/channelz_registry_test.cc [hunk 4]
```diff
@@ -43,56 +43,151 @@ namespace grpc_core {
...
+      new BaseNode(BaseNode::EntityType::kTopLevelChannel));
```
Comment (markdroth) [inline test/core/channel/channelz_registry_test.cc]: Need to use `New<>`.  (This problem goes away if you just use `MakeUnique<>`.)

File: test/core/channel/channelz_registry_test.cc (hunk 3 above)
Comment (ncteisen) [reply test/core/channel/channelz_registry_test.cc]: Done

File: test/core/channel/channelz_registry_test.cc (hunk 4 above)
Comment (ncteisen) [reply test/core/channel/channelz_registry_test.cc]: Done

File: src/core/lib/gprpp/inlined_vector.h [hunk 5]
```diff
@@ -123,6 +123,14 @@ class InlinedVector {
 
   void push_back(T&& value) { emplace_back(std::move(value)); }
 
+  void pop_back() {
```
[the remark this answers is not in the document]
Comment (ncteisen) [reply src/core/lib/gprpp/inlined_vector.h]: Will do this this afternoon

**Comment by ncteisen** [comment]
Addressed all but the test for pop_back. Will send that commit out after lunch

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                               FILE SIZE
 ++++++++++++++ GROWING                                                                 ++++++++++++++
  +0.0%    +312 [None]                                                                  +3.26Ki  +0.0%
   +20%    +512 src/core/lib/channel/channelz_registry.cc                                  +512   +20%
      [NEW]    +214 grpc_core::channelz::ChannelzRegistry::MaybePerformCompactionLocked        +214  [NEW]
      [NEW]    +153 grpc_core::channelz::ChannelzRegistry::FindByUuidLocked                    +153  [NEW]
       +47%     +80 grpc_core::channelz::ChannelzRegistry::InternalUnregister                   +80   +47%
       +55%     +21 grpc_core::channelz::ChannelzRegistry::ChannelzRegistry                     +21   +55%
      +7.4%     +16 grpc_core::channelz::ChannelzRegistry::InternalRegister                     +16  +7.4%
      +9.9%     +15 [Unmapped]                                                                  +15  +9.9%
       +15%     +13 grpc_core::channelz::ChannelzRegistry::InternalGet                          +13   +15%

  +0.1%    +824 TOTAL                                                                   +3.76Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +24  +0.0%

  [ = ]       0 TOTAL      +24  +0.0%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,965,924      Total (>)      1,965,335

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,888,757      Total (>)     10,888,231

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,965,924      Total (>)      1,965,335

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,888,755      Total (>)     10,888,227

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                               FILE SIZE
 ++++++++++++++ GROWING                                                                 ++++++++++++++
  +0.0%    +312 [None]                                                                  +3.27Ki  +0.0%
   +20%    +512 src/core/lib/channel/channelz_registry.cc                                  +512   +20%
      [NEW]    +214 grpc_core::channelz::ChannelzRegistry::MaybePerformCompactionLocked        +214  [NEW]
      [NEW]    +153 grpc_core::channelz::ChannelzRegistry::FindByUuidLocked                    +153  [NEW]
       +47%     +80 grpc_core::channelz::ChannelzRegistry::InternalUnregister                   +80   +47%
       +55%     +21 grpc_core::channelz::ChannelzRegistry::ChannelzRegistry                     +21   +55%
      +7.4%     +16 grpc_core::channelz::ChannelzRegistry::InternalRegister                     +16  +7.4%
      +9.9%     +15 [Unmapped]                                                                  +15  +9.9%
       +15%     +13 grpc_core::channelz::ChannelzRegistry::InternalGet                          +13   +15%

  +0.1%    +824 TOTAL                                                                   +3.77Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +24  +0.0%

  [ = ]       0 TOTAL      +24  +0.0%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,965,924      Total (>)      1,965,335

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,888,755      Total (>)     10,888,224

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: src/core/lib/channel/channelz_registry.cc [hunk 6]
```diff
@@ -56,23 +56,71 @@ ChannelzRegistry::~ChannelzRegistry() { gpr_mu_destroy(&mu_); }
...
+      double(num_empty_slots_) / double(entities_.capacity());
+  if (emptiness_ratio > kEmptinessTheshold) {
+    int front = 0;
+    for (size_t i = 0; i < entities_.size(); ++i) {
+      if (entities_[i] != nullptr) {
+        entities_[front++] = entities_[i];
+      }
+    }
+    for (int i = 0; i < num_empty_slots_; ++i) {
+      entities_.pop_back();
```
Comment (markdroth) [inline src/core/lib/channel/channelz_registry.cc]: I just noticed that `pop_back()` doesn't actually shrink the capacity of the `InlinedVector`.  So the capacity will grow based on the maximum number of elements ever seen, and will never shrink again.

This is probably okay for now, but I suspect we're going to want to replace this with a map at some point in the future.

**Pull request merged**